src/motion/space_groups.F

```diff
@@ -0,0 +1,1229 @@
+!> \param spgr ...
+!> \param particle_set ...
+!> \param cell ...
+!> \param geo_section ...
+!> \param gopt_env ...
+!> \param iounit ...
+!> \par History
+!>      01.2020 created [pcazade]
+!> \author Pierre-André Cazade (first version)
+!> \note  rotation matrices innclude translations and translation symmetry:
+!>        it works with supercells as well.
+! **************************************************************************************************
+   SUBROUTINE identify_space_group(spgr,subsys, cell, geo_section, gopt_env, iounit)
+    
+      IMPLICIT NONE
+      
+      TYPE(section_vals_type), POINTER                   :: geo_section
+      TYPE(gopt_f_type), POINTER                         :: gopt_env
+      !TYPE(particle_type), DIMENSION(:), POINTER         :: particle_set
+      TYPE(cp_subsys_type), POINTER                      :: subsys
+      TYPE(cell_type), POINTER                           :: cell
+      TYPE(particle_list_type), POINTER                  :: particles, &
+                                                            shell_particles, &
+                                                            core_particles
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'identify_space_group', &
+         routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: handle, i, iounit, nparticle
+      INTEGER                                            :: n_sr_rep, n_atom, n_core
+      INTEGER                                            :: k, shell_index, n_shell
+      INTEGER, DIMENSION(:), POINTER                     :: tmp
+      INTEGER, ALLOCATABLE, DIMENSION(:)                 :: atype
+      INTEGER, ALLOCATABLE, DIMENSION(:,:)               :: ranges
+      REAL(KIND=dp)                                      :: eps_geo, eps_sym
+      REAL(KIND=dp), DIMENSION(:), POINTER               :: pol
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:, :)        :: scoord
+      TYPE(spgr_type)                                    :: spgr
+
+      CALL timeset(routineN, handle)
+
+      n_sr_rep = 0
+      nparticle = 0
+      n_atom = 0
+      n_core = 0
+      n_shell = 0
+
+      NULLIFY (particles)
+      NULLIFY (core_particles)
+      NULLIFY (shell_particles)
+
+      particles => subsys%particles
+      shell_particles => subsys%shell_particles
+      core_particles => subsys%core_particles
+
+      CPASSERT(ASSOCIATED(particles))
+      n_atom = particles%n_els
+      ! Check if we have other kinds of particles in this subsystem
+      IF (ASSOCIATED(shell_particles)) THEN
+         n_shell = shell_particles%n_els
+         CPASSERT(ASSOCIATED(core_particles))
+         n_core = subsys%core_particles%n_els
+         ! The same number of shell and core particles is assumed
+         CPASSERT(n_core == n_shell)
+      ELSE IF (ASSOCIATED(core_particles)) THEN
+         ! This case should not occur at the moment
+         CPASSERT(ASSOCIATED(shell_particles))
```

Comment:
does this assertion trigger at all? If `ASSOCIATED(shell_particles)` is `.TRUE.` we would be in the first `IF` branch, right?